source/opt/fix_func_call_arguments.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2022 AMD LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fix_func_call_arguments.h"
+
+#include "ir_builder.h"
+
+using namespace spvtools;
+using namespace opt;
+
+bool FixFuncCallArgumentsPass::HasNoFunctionToCall() {
+  auto funcsNum = get_module()->end() - get_module()->begin();
+  return funcsNum == 1;
+}
+
+Pass::Status FixFuncCallArgumentsPass::Process() {
+  bool modified = false;
+  if (HasNoFunctionToCall()) return Status::SuccessWithoutChange;
+  for (auto& func : *get_module()) {
+    // Get Variable insertion point
+    Instruction* varInsertPt = &*(func.begin()->begin());
+    func.ForEachInst([this, &modified, &varInsertPt](Instruction* inst) {
+      if (inst->opcode() == SpvOpFunctionCall) {
+        modified |= FixFuncCallArguments(inst, varInsertPt);
+      }
+    });
+  }
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool FixFuncCallArgumentsPass::FixFuncCallArguments(Instruction* func_call_inst,
+                                                    Instruction* var_insertPt) {
+  bool modified = false;
+  Instruction* nextInst = func_call_inst->NextNode();
+  for (uint32_t i = 0; i < func_call_inst->NumInOperands(); ++i) {
+    Operand& op = func_call_inst->GetInOperand(i);
+    if (op.type != SPV_OPERAND_TYPE_ID) continue;
+    Instruction* operand_inst = get_def_use_mgr()->GetDef(op.AsId());
+    if (operand_inst->opcode() == SpvOpAccessChain) {
+      ReplaceAccessChainFuncCallArguments(func_call_inst, &op, operand_inst,
+                                          nextInst, var_insertPt, i);
+      modified = true;
+    }
+  }
+  if (modified) {
+    context()->UpdateDefUse(func_call_inst);
+  }
+  return modified;
+}
+
+void FixFuncCallArgumentsPass::ReplaceAccessChainFuncCallArguments(
+    Instruction* func_call_inst, Operand* operand, Instruction* operand_inst,
+    Instruction* next_insertPt, Instruction* var_insertPt,
+    unsigned operand_index) {
```

Comment:
This function has a lot of parameters.  It would make it easier to understand and use if there were fewer.  Here are a few suggestions:

1) `operand` is only used to get the id right?  If so, you can get the id from the result id of operand_inst.
2) Could you do `Instruction* nextInst = func_call_inst->NextNode();` in this function instead of in its caller?  I don't think it will save anything loading outside the loop since it is just a single load.
3) Could you have this function return the id of the new variable, and then do `func_call_inst->SetInOperand(operand_index, {var->result_id()});` in the caller?  That way `operand_index` does not have to be a parameter.
4) You can do `context()->get_instr_block(func_call_inst)->GetParent()` to get the function, and then you can compute `var_insertPt` from that.

Reply:
the insert point should passed from out of loop, so we can keep insertion order, first inserted variable for the first operand.

Reply:
I don't think it's that important.  I would prefer more read able C++.